File: frontend/src/lib/connectors.ts
/**
 * The `/connectors` contract, mirrored from `src/schemas/connectors.py`.
 *
 * The point of `ConnectorField` is that the panel renders a form from data
 * rather than from a component per service. A connector added to the backend
 * registry shows up here with its own inputs, correctly labelled and correctly
 * masked, without this file or the panel changing.
 *
 * Credentials travel one way. `connect()` sends them; nothing returns one.
 * `hints` carries the non-secret fields as typed plus the last four characters
 * of the secret one, which is what lets a row read "vec-chunks · ····8fa2".
 */

export type ConnectorField = {
  name: string;
  label: string;
  /** Renders as a password and is never echoed back. */
  secret: boolean;
  required: boolean;
  placeholder: string;
  help: string;
};

/** "tools" is Composio; "vector" is where a user's embeddings live. */
export type ConnectorKind = "tools" | "vector";

export type Connector = {
  slug: string;
  name: string;
  kind: ConnectorKind;
  summary: string;
  docsUrl: string;
  fields: ConnectorField[];

  connected: boolean;
  hints: Record<string, string>;
  connectedAt: string | null;
  updatedAt: string | null;
  /** Stored but no longer decrypting — the master key was rotated. */
  stale: boolean;
};

export type ConnectorList = {
  connectors: Connector[];
  /** False when the server cannot hold a credential at all. */
  configured: boolean;
  /** Which store answers this user's retrieval; null = the deployment's own. */
  vectorBackend: string | null;
};

export class ConnectorError extends Error {
  readonly status: number;

  constructor(message: string, status: number) {
    super(message);
    this.name = "ConnectorError";
    this.status = status;
  }

  get signedOut(): boolean {
    return this.status === 401;
  }
}

async function fail(response: Response): Promise<never> {
  let detail = "Something went wrong.";
  try {
    const body = (await response.json()) as { detail?: string };
    if (typeof body.detail === "string") detail = body.detail;
  } catch {
    // no JSON body — keep the generic message
  }
  throw new ConnectorError(detail, response.status);
}

/**
 * Every connector and this account's state on each.
 *
 * Empty-but-unconfigured on failure rather than thrown: this feeds a panel
 * that opens on a click, and it should read as unavailable rather than take
 * the page down.
 */
export async function listConnectors(signal?: AbortSignal): Promise<ConnectorList> {
  try {
    const response = await fetch("/api/connectors", { signal, cache: "no-store" });
    if (!response.ok) return { connectors: [], configured: false, vectorBackend: null };

    return (await response.json()) as ConnectorList;
  } catch {
    return { connectors: [], configured: false, vectorBackend: null };
  }
}

/** Attach a service. Throws — somebody filled in a form and pressed a button. */
export async function connect(
  slug: string,
  values: Record<string, string>,
): Promise<Connector> {
  const response = await fetch(`/api/connectors/${encodeURIComponent(slug)}`, {
    method: "PUT",
    headers: { "content-type": "application/json" },
    body: JSON.stringify({ values }),
  });

  if (!response.ok) return fail(response);
  return (await response.json()) as Connector;
}

export async function disconnect(slug: string): Promise<void> {
  const response = await fetch(`/api/connectors/${encodeURIComponent(slug)}`, {
    method: "DELETE",
  });
  if (!response.ok) await fail(response);
}

/** What a row shows under its name: the readable half of what was stored. */
export function describeHints(connector: Connector): string {
  const parts = Object.entries(connector.hints)
    .filter(([, value]) => value)
    .map(([key, value]) => (key.endsWith("_hint") ? `····${value}` : value));

  return parts.join(" · ");
}

File: src/schemas/connectors.py
"""The `/connectors` contract — enough for the panel to render a form it has
never seen.

`ConnectorField` is the interesting shape. It is a *description of an input*,
sent to the browser so the panel can build the right form for a connector added
after that panel was written. Without it, every new connector means a React
component, and the form and the server-side validation drift the first time
either is edited.

**Credentials travel one way.** They go in on `ConnectCredentials` and nothing
sends one back. `hints` carries the non-secret fields as typed plus the last
four characters of the secret one, which is what lets the panel say
"Pinecone · vec-chunks · ····8fa2" without ever holding the key.
"""

from datetime import datetime

from pydantic import Field

from src.schemas.wire import Wire


class ConnectorField(Wire):
    """One input the panel should render."""

    name: str
    label: str
    # Renders as a password, is never echoed back, and only its last four
    # characters are kept readable. The single most important flag here.
    secret: bool = False
    required: bool = True
    placeholder: str = ""
    help: str = ""


class Connector(Wire):
    """One connector, described, with this user's state on it."""

    slug: str
    name: str
    # "tools" or "vector". The panel groups by it, and retrieval asks for it
    # rather than enumerating slugs.
    kind: str
    summary: str
    docs_url: str = ""
    fields: list[ConnectorField] = []

    connected: bool = False
    hints: dict[str, str] = {}
    connected_at: datetime | None = None
    updated_at: datetime | None = None
    # A credential is stored but no longer decrypts — the master key was
    # rotated. "Connected" would be a lie and "not connected" loses the fact
    # that there is a row to replace, so it is its own flag.
    stale: bool = False


class ConnectorList(Wire):
    connectors: list[Connector] = []
    # False when the server has no COMPOSIO_ENCRYPTION_KEY and so cannot hold a
    # credential at all. Different from "you have not connected anything", and
    # only one of the two is the signed-in user's to fix.
    configured: bool = True
    # Which connector is currently answering this user's retrieval. Null means
    # *nothing* answers it — there is no deployment corpus behind this app, so
    # a user with no vector store attached has nowhere to be searched and /ask
    # says so (docs/13-connectors.md).
    vector_backend: str | None = None


class ProfileField(Wire):
    """One metadata key on the connected store, and how much of it carries it.

    `coverage` rather than a bare presence flag: a key on 3% of an index is not
    something a query may filter on, and the two are indistinguishable to
    anything that only records whether the key was ever seen.
    """

    name: str
    types: list[str] = []
    coverage: float = 0.0
    distinct: int | None = None
    examples: list[str] = []
    filterable: bool = False
    # One value across the whole sample — carried, and useless for narrowing.
    constant: bool = False


class ConnectorProfile(Wire):
    """What this app understood about a connected store, measured and written.

    Two halves that are deliberately not merged. `facts` is measured and is
    what the retrieval path acts on; `summary`, `topics`, `good_for` and
    `not_for` are written by a model over sampled text and only steer routing.
    A caller that confuses them turns a wrong sentence into a wrong query.
    """

    connector: str
    kind: str
    # pending | ok | degraded | failed
    status: str
    # Where it points, with no credential in it.
    location: str = ""
    reachable: bool = False

    title: str = ""
    summary: str = ""
    topics: list[str] = []
    good_for: list[str] = []
    not_for: list[str] = []

    records: int | None = None
    dimensions: int | None = None
    metric: str = ""
    index: str = ""
    sampled: int = 0
    text_field: str = ""
    scripts: list[str] = []
    fields: list[ProfileField] = []
    # Things a person should be told in their own words — a column that is the
    # same on every row, a namespace holding vectors nobody is searching.
    notes: list[str] = []

    # The measured capabilities. `filters` here is the answer every backend
    # used to hard-code as true.
    lexical: bool = False
    filters: bool = False
    parallel_text: bool = False
    searchable: bool = False

    # The agent-facing rendering: what goes into a system prompt verbatim.
    card: str = ""
    error: str = ""
    profiled_at: datetime | None = None


class Capabilities(Wire):
    """Everything the agent knows it can reach right now, in one call.

    `card` is the block that goes into a system prompt as-is. The structured
    profiles are beside it for a panel that wants to render the same thing with
    its own layout, never as a second source of truth.
    """

    card: str = ""
    profiles: list[ConnectorProfile] = []
    # Which store is answering retrieval. Null means nothing is: retrieval
    # abstains until this user connects one.
    vector_backend: str | None = None


class ConnectCredentials(Wire):
    """The one request on this API that carries secrets.

    Free-form because the fields differ per connector; the server takes only
    what that connector declared and drops the rest, so this being open does
    not mean arbitrary JSON reaches a credential blob.

    No `user_id`: whose account this becomes is decided by the verified Clerk
    token, and accepting one here would let anybody file a key under somebody
    else's name.
    """

    values: dict[str, str] = Field(
        default_factory=dict, description="Field name → value, per the connector's schema"
    )


class Disconnected(Wire):
    connector: str
    removed: bool
